src/index.ts

```diff
@@ -589,6 +561,61 @@ class Envoy {
       return new Package(true, { "type": "success" }, { response_id });
     }, this.opts.NS);
   }
+
+  async signIn(hha_hash, agent_id): Promise<void> {
+    if (agent_id in this.anonymous_agents) {
+      // Nothing to do. Anonymous cell is always active
+      return;
+    }
+
+    const hosted_agent_instance_app_id = `${hha_hash}:${agent_id}`;
+
+    // Activate App - Tell Holochain to begin gossiping and be ready for zome calls on this app.
+    try {
+      log.info("Activating Installed App (%s)", hosted_agent_instance_app_id);
+      const adminResponse = await this.callConductor("admin", 'activateApp', { installed_app_id: hosted_agent_instance_app_id });
+
+      if (adminResponse.type !== "success") {
+        log.error("Conductor 'activateApp' returned non-success response: %s", adminResponse);
+        throw (new HoloError(`Failed to complete 'activateApp' for installed_app_id'${hosted_agent_instance_app_id}'.`)).toJSON();
+      }
+    } catch (err) {
+      if (err.message.includes("Tried to activate an app that was not installed")) {
+        // This error is returned in two cases:
+        // a) The app is not installed -- Return an error to the user saying that they may need to sign up first.
+        // b) The app is already activated -- Our job is done.
+
+        // Check for the second case using appInfo
+        try {
+          const appInfo = await this.callConductor("app", { installed_app_id: hosted_agent_instance_app_id });
+          // Check that the appInfo result was not null (would indicate app not installed)
+          if (appInfo.installed_app_id !== undefined) {
+            log.normal("Completed sign-in process for Agent (%s) HHA ID (%s)", agent_id, hha_hash);
+            return;
+          }
+        } catch (appInfoErr) {
+          log.error("Failed during 'appInfo': %s", String(appInfoErr));
+          throw (new HoloError(`Failed to complete 'appInfo' for installed_app_id'${hosted_agent_instance_app_id}'.`)).toJSON();
+        }
+      }
+      log.error("Failed during 'activateApp': %s", String(err));
+      throw err;
+    }
+  }
+
+  async signOut(agent_id): Promise<void> {
+    // Assuming agent is not anonymous, we need to deactivate all their hApps.
+
+    const regex = new RegExp(`:${agent_id}$`);
+
+    const activeApps = await this.callConductor("admin", "listActiveApps");
+    await Promise.all(activeApps.map(async (installed_app_id: string) => {
+      if (regex.test(installed_app_id)) {
```

Comment:
This would deactivate all the happs that were active, by an agent. This would only work if we are under the assumption that every agent signs into one app at a time. Or that every app an agent signs into they have a different `agent_id`

Reply:
Good point. My proposal would be to continue deactivating all hApps of that agent ID upon wormhole timeout, since that agent can't fulfill signing requests for any hApp, but modify the on("close") handler so that it only deactivates the specific hApp that was involved with that connection